rids/network_capture.py

```diff
@@ -14,74 +14,103 @@
 
 
 import ipaddress
-import logging
+import subprocess
 
+from rids import observations
 
-allowed_sni_port = set([
-    ('mtalk.google.com', 5228),
-    ('proxy-safebrowsing.googleapis.com', 80),
-    ('courier.push.apple.com', 5223),
-    ('imap.gmail.com', 993),
-])
 
+class IpPacketMonitor:
+  """Continuously scans the network traffic for remote IPs being contacted.
+  The scan() function is an iterator that produces a dict{...} representing
+  the endpoints (specifically the `remote_ip` and `remote_port` fields).  This
+  scanning will continue until the pipe from the underlying network scanning
+  process is closed.
+  """
+  def __init__(self, host_ip):
+    self._host_ip = host_ip
 
-# TODO: decouple detection and logging
+  def monitor(self):
+    """Generator for observations of remote IP addresses."""
+    proc = _start_tshark([
+        '-f', f'ip and src ip == {self._host_ip}',
+        '-Tfields',
+        # The order of the following fields determines the ordering in output
+        '-e', 'frame.time',
+        '-e', 'ip.dst',
+        '-l',
+    ])
 
-def detect_tls_events(input_stream):
-    unusual_tls_traffic = {}  # map[int]dict stream_id -> connection_details
+    for line in iter(proc.stdout.readline, b''):
+      values = line.split('\t')
+      ip_info = observations.IpPacket(
+        timestamp=values[0],
+        remote_ip=ipaddress.ip_address(values[1]))
+      yield ip_info
 
-    logging.basicConfig(level=logging.DEBUG,
-                        format='%(asctime)s %(levelname)-8s %(message)s',
-                        datefmt='%m-%d %H:%M',
-                        filename='/var/maldetector.log',
-                        filemode='a')
 
-    for line in iter(input_stream, b''):
-        values = line.split('\t')
-        if len(values) >= 8:
-            stream_id = int(values[1])
-            if int(values[2]) == 1:  # client hello
-                sni_and_port = (values[7], int(values[6]))
-                if sni_and_port not in allowed_sni_port:
-                    unusual_tls_traffic[stream_id] = {
-                        'client_ts': values[0],
-                        'server_ip': values[5],
-                        'server_port': values[6],
-                        'server_name': values[7],
-                        'ja3': values[8],
-                        'ja3_full': values[9],
-                    }
-            elif int(values[2]) == 2:  # server hello
-                port = int(values[4])
-                if port == 443:
-                    continue
-                connection_details = unusual_tls_traffic.get(stream_id, None)
-                if not connection_details:
-                    continue
-                connection_details['ja3s'] = values[8]
-                connection_details['ja3s_full'] = values[9]
-                logging.info('connection_details %s', connection_details)
-                del unusual_tls_traffic[stream_id]
+class TlsConnectionMonitor:
```

Comment:
I'd put these in different files, since they are unrelated and it's easier to understand them in isolation. It would also make sense to separate the IpPacket and TlsConnection types too. Perhaps we put them together with the class the generates them. So instead of `network_capture.py` and `observations.py` you have `ip_monitor.py` and `tls_monitor.py`. That's a more useful boundary.

You can put the `IpRule` in the ip_monitor file too, so we can fully understand it in one place. It's all part of the Ip monitor interface.

Reply:
yeah I also had started to think we would want multiple modules replacing this network_capture module, especially as I was adding the third case for domain monitoring.  We can pull out the common tshark calling code and have a separate model for each group of network features.  This means IOC parsing now depends on the network models for the rule types it uses, and I kind of liked it better when it just depended on the rule definitions, but I appreciate the benefits of slicing vertically too.